### PR TITLE
Enabling Unclean Session Publish Re-Transmits

### PR DESCRIPTION
**Pull request by DakshitBabbar**
Description
-----------

This PR enables the coreMQTT library to resend unacked publishes on an unclean session connection.

Following is a brief summary of changes:
1. Add a new API `MQTT_InitRetransmits` that will initialise the context to handle publish retransmits on an unclean session connection
2. Add signatures of callback function pointers that users will define in order to:
     a. copy and store outgoing publishes
     b. retrieve copied publish on an unclean session connection to resend
     c. clear a copied publish when a `PUBACK`/`PUBREC` is received
     d. clear all copied publishes on a clean session connection
3. Update the API's to check if callback's are defined and implement resend publishes as required.

Following are the specifics of the changes:
1. Add 3 new MQTTStatus_t values: MQTTPublishStoreFailed, MQTTPublishRetrieveFailed and MQTTPublishClearAllFailed
2. Update `MQTTContext_t` to hold the callback function pointers
        a. `MQTTRetransmitStorePacket storeFunction`
        b. `MQTTRetransmitRetrievePacket retrieveFunction`
        c. `MQTTRetransmitClearPacket clearFunction`
        d. `MQTTRetransmitClearAllPackets clearAllFunction`
3. Update the `MQTT_Status_strerror`  function to handle the new `MQTTStatus_t` values
4. Add a new API function `MQTT_InitRetransmits` that will initialise the new callback functions in the `MQTTContext_t`
5. Add this API to the core_mqtt.h file to make it available to users
6. Modify `MQTT_Publish`
        a. copy the outgoing publish packet in form of an array of `TransportOutVector_t` if the callback if defined
        b. if copy fails then bubble up  corresponding error status code
7. Modify `MQTT_ReceiveLoop`
        a. on receiving a `PUBACK`/`PUBREC` clear the copy of that particular publish after the state of the publish record has been successfully updated, if the callback if defined
8. Modify `MQTT_Connect`
        a. on a clean session clear all the copies of publishes stored if the callback is defined
        b. if clear all fails then bubble up  corresponding error status code
        c. on an unclean session get the packetID of the unacked publishes and retrieve the copies of those if the callback is defined
        d. if retrieve fails then bubble up  corresponding error status code

Approaches Taken
---------------
- To let user know about the changes we have made we will add them to a changelog and have a minor version bump
- To be in line with the zero copy principle in our library we chose to provide and retrieve the publish packets for storing and resending in form of an array of `TransportOutVector_t`
- Code is written in a way that on receiving a `PUBACK`/`PUBREC` the copy will be cleared after the state of the publish record is changed so that if state update fails the copy won't be cleared. Otherwise if the state does not change and the copy is cleared then when a connection is made with an unclean session there will be a retrieve fail as the system is in an inconsistent state.
- We are storing the copies of the publishes with the Duplicate flag set this is because on retrieving the packet we will get it in the form of a `TransportOutVector_t` that holds the data in a `const` pointer which cannot be changed after retrieving.

Pending Tasks
---------------
- [ ] Changelog
- [ ] Minor version bump
- [x] Doxygen example for the new API
- [x] Better API Names
- [x] Unit Test Updates
- [x] CBMC Proof

### CODE REVIEW

File: source/include/core_mqtt.h [hunk 1]
```diff
@@ -101,6 +101,68 @@ typedef void (* MQTTEventCallback_t )( struct MQTTContext * pContext,
...
+ *
+ * @param[in] pContext Initialised MQTT Context.
+ * @param[in] packetId Copied publish packet identifier.
+ * @param[in] pIoVec Output parameter to store the pointer to the copied publish packet form of array of Tansport Vectors.
+ * @param[in] ioVecCount Output parameter to store the number of transport vectors in the pIoVec array.
+ *
+ * @return True if the retreive is successful else false.
+ */
+/* @[define_mqtt_retransmitretrievepacket] */
+typedef bool ( * MQTTRetransmitRetrievePacket)( struct MQTTContext * pContext,
```
[the remark this answers is not in the document]
Comment (DakshitBabbar) [reply source/include/core_mqtt.h]: I feel that, provided we are giving them the `TransportOutVector_t` array to copy while storing, if we make them store the packet in a contiguous memory then it will be more difficult for them to manage their heap. This is because:
1. it will have untidy way to search for a packet ID during retrieval if they store the packet ID within the same memory.
2. while clearing a packet they will have to either leave an empty block which is not guaranteed to be fit for storing a new packet or they will have to move the other packets back to fill in the empty space (which is only possible if the packet ID is stored within the same memory else we cannot move the data is it will tamper with the pointers that are meant to store the data)
3. as can be seen from point 2, big chunks of fragmented memory might still remain.

If stored in the form of `TransportOutVector_t` array itself, it will be easier for them to manage their heap. Easy search, easy storing, tidy clearing.

Hence, even though both the methods lead to similar levels of fragmentation, the method of storing in form of `TransportOutVector_t` array is tidier. 

I have created an example for the way the users can write these callbacks in a tidy manner.

```
#define MAX_PUBLISH_COUNT 3U

#define MAX_SIZE_OF_PUBLISH_PACKET 64U

#define MAX_PUBLISH_VECTOR_COUNT 4U

typedef struct publishCopy
{
    uint16_t packetId;

    uint8_t buffer[ MAX_SIZE_OF_PUBLISH_PACKET ];

    TransportOutVector_t pIoVec[ MAX_PUBLISH_VECTOR_COUNT ];

} publishCopy_t;

static publishCopy_t publishCopyBuffer[ MAX_PUBLISH_COUNT ];

bool publishStoreCallback(struct MQTTContext* pContext,
                        uint16_t packetId,
                        TransportOutVector_t* pIoVec,
                        size_t ioVecCount)
{
    bool found = false;

    for ( int i = 0; i < MAX_PUBLISH_COUNT; i++ )
    {
        if (publishCopyBuffer[i].packetId == 0)
        {
            uint8_t* bufferPointer = publishCopyBuffer[i].buffer;

            publishCopyBuffer[i].packetId = packetId;

            for (int j = 0; j < ioVecCount; j++)
            {
                memcpy( ( void* ) bufferPointer, ( const void* ) pIoVec[j].iov_base, pIoVec[j].iov_len);

                publishCopyBuffer[i].pIoVec[j].iov_base = bufferPointer;
                publishCopyBuffer[i].pIoVec[j].iov_len = pIoVec[j].iov_len;

                bufferPointer += pIoVec[j].iov_len;
            }

            found = true;

            break;
        }
    }

    return found;
}

bool publishRetrieveCallback(struct MQTTContext* pContext,
                            uint16_t packetId,
                            TransportOutVector_t** pIoVec,
                            size_t* ioVecCount)
{
    bool found = false;

    for ( int i = 0; i < MAX_PUBLISH_COUNT; i++ )
    {
        if (publishCopyBuffer[i].packetId == packetId)
        {
            *pIoVec = publishCopyBuffer[i].pIoVec;
            *ioVecCount = 4;

            found = true;

            break;
        }
    }

    return found;
}

bool publishClearCallback(struct MQTTContext* pContext,
                          uint16_t packetId)
{
    bool found = false;

    for (int i = 0; i < MAX_PUBLISH_COUNT; i++)
    {
        if (publishCopyBuffer[i].packetId == packetId)
        {
            memset(&publishCopyBuffer[i], 0, sizeof(publishCopy_t));

            found = true;

            break;
        }
    }

    return found;
}
```

File: source/include/core_mqtt_serializer.h [hunk 2]
```diff
@@ -85,23 +85,30 @@ struct MQTTPacketInfo;
...
+                                    a delay). */
+    MQTTStatusConnected,            /**< MQTT connection is established with the broker */
+    MQTTStatusNotConnected,         /**< MQTT connection is not established with the broker */
+    MQTTStatusDisconnectPending,    /**< Transport Interface has failed and MQTT connection needs to be closed */
+    MQTTPublishStoreFailed,         /**< User provided API to store a copy of outgoing publish for retransmission  purposes, 
+                                    has failed */
+    MQTTPublishRetrieveFailed,      /**< User provided API to retrieve the copy of a publish while reconnecting 
+                                    with an unclean session has failed */
+    MQTTPublishClearFailed,         /**< User provided API to clear the copy of a publish on receiving a PUBACK has failed */
+    MQTTPublishClearAllFailed       /**< User provided API to clear all the copies of publishes while connecting with a clean
```
[the remark this answers is not in the document]
Comment (DakshitBabbar) [reply source/include/core_mqtt_serializer.h]: Yes I think we should, we should do this for issue#1 as well.

File: source/core_mqtt_serializer.c [hunk 3]
```diff
@@ -2623,6 +2628,32 @@ MQTTStatus_t MQTT_GetIncomingPacketTypeAndLength( TransportRecv_t readFunc,
...
+
+    if(pHeader == NULL)
+    {
+        status = MQTTBadParameter;
+    } 
+    else if( ( ( *pHeader ) & MQTT_PACKET_TYPE_PUBLISH ) == 0 )
+    {
+        status = MQTTBadParameter;
+    } 
+    else if( set )
```
[the remark this answers is not in the document]
Comment (DakshitBabbar) [reply source/core_mqtt_serializer.c:2644]: Just wanted to know if this is a MISRA violation specific change or in general `set == true` if preffered more? because according to me, if the variable within the if condition is of type boolean then comparing it with `true` or `false` is less efficient as compared to using it directly.

File: source/core_mqtt.c [hunk 4]
```diff
@@ -2190,7 +2204,30 @@ static MQTTStatus_t sendPublishWithoutCopy( MQTTContext_t * pContext,
         totalMessageLength += pPublishInfo->payloadLength;
     }
 
-    if( sendMessageVector( pContext, pIoVector, ioVectorLength ) != ( int32_t ) totalMessageLength )
+    /* store a copy of the publish with duplicate flag set for retransmission purposes */
```
[the remark this answers is not in the document]
Comment (DakshitBabbar) [reply source/core_mqtt.c]: done

File: source/include/core_mqtt.h [hunk 5]
```diff
@@ -101,6 +101,68 @@ typedef void (* MQTTEventCallback_t )( struct MQTTContext * pContext,
...
+/* @[define_mqtt_retransmitstorepacket] */
+
+/**
+ * @brief User defined API used to retreive a copied publish for resend operation. Used to 
+ * track any publish retransmit on an unclean session connection.
+ *
+ * @param[in] pContext Initialised MQTT Context.
+ * @param[in] packetId Copied publish packet identifier.
+ * @param[in] pIoVec Output parameter to store the pointer to the copied publish packet form of array of Tansport Vectors.
+ * @param[in] ioVecCount Output parameter to store the number of transport vectors in the pIoVec array.
```
[the remark this answers is not in the document]
Comment (DakshitBabbar) [reply source/include/core_mqtt.h:129]: done

File: source/core_mqtt_serializer.c [hunk 6]
```diff
@@ -101,6 +101,11 @@
  */
 #define UINT8_SET_BIT( x, position )      ( ( x ) = ( uint8_t ) ( ( x ) | ( 0x01U << ( position ) ) ) )
 
+/**
+ * @brief Set a bit in an 8-bit unsigned integer.
```
[the remark this answers is not in the document]
Comment (DakshitBabbar) [reply source/core_mqtt_serializer.c]: done

File: source/include/core_mqtt_serializer.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (DakshitBabbar) [reply source/include/core_mqtt_serializer.h]: done for both

**Review by archigup — APPROVED** [review]

File: source/core_mqtt_serializer.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (archigup) [reply source/core_mqtt_serializer.c]: Both will compile down to the same thing. Misra allows either as long as set is 'essentially boolean' as is defined in one of the appendixes. For pointers, `== NULL` is required as the if condition expression must be essentially boolean but a pointer isnt. The result of `==` is though.

[the remark this answers is not in the document]
Comment (AniruddhaKanhere) [reply source/core_mqtt_serializer.c]: Agree with Archit.

`if(set == true) and `if(set)` both lead to the compiler generating same code because the latter is intrinsic check against being true or false and the previous is an extrinsic check. They make our life easier - not the compilers :) 

But, for  the sake of consistency, we should make it as I suggested.

[the remark this answers is not in the document]
Comment (archigup) [reply source/core_mqtt_serializer.c]: Well `if(x)` semantically jumps over the section if x is zero. but compiler will see that `== true` is not changing the result and will drop it.

[the remark this answers is not in the document]
Comment (archigup) [reply source/core_mqtt_serializer.c]: for the iovec thing, I think it makes sense as allows more flexibility in storage. I'd probably use contiguous memory, but can do that by just passing a single elem iovec on return and concatenating the elems on receiving.

File: source/include/core_mqtt.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (archigup) [reply source/include/core_mqtt.h]: I think it makes sense as allows more flexibility in storage. I'd probably use contiguous memory, but can do that by just passing a single elem iovec on return and concatenating the elems on receiving.

Though if nobody would use the flexibility in storage, its just additional overhead for all users to deal with.
For saving a packet, I can see it being useful though as the MQTT library should not need the memory for concatenating the segments; this way the user can concat them directly into the destination buffer. For retrieving a packet, I'm not sure if I see much value other than consistency.

[the remark this answers is not in the document]
Comment (AniruddhaKanhere) [reply source/include/core_mqtt.h]: Thank you for the answer @DakshitBabbar,

The method suggested above seems a bit inefficient in terms of memory. What I mean by that is the user must allocate arrays of fixed size. That leads to inefficient use of memory which is quite an important resource on embedded systems. Any packet which is even 2 bytes long will take up whole slot which can used by others.

For an example, let us consider a scenario where the outstanding packet is limited to 2 and the array size is 1000. So, 2 packets of size 500 and 1500 cannot be sent. Even though there is memory.

Also, the method suggested above has more memory requirement only to store the data-structure. It is storing array of arrays.

Also, when the user just wants to call malloc/free to store and clear a value they will need to call malloc and free on many vectors one at a time. It will cause very small memory allocations and deallocations which causes fragmentation.

What I am proposing is this:

```
typedef struct userPublishStore {
    uint8_t *buffer;
    size_t bufferlen;
    uint16_t packetId;
} userPublishStore_t;

userPublishStore_t storedPublishes[MAX_PUBLISH_COUNT];

bool publishStoreCallback(
    struct MQTTContext *pContext,
    uint16_t packetId,
    TransportOutVector_t *pIoVec,
    size_t ioVecCount
) {
    bool ret = false;
    size_t memory_required = 0;
    size_t store_location;

    for (store_location = 0; store_location < MAX_PUBLISH_COUNT;
         store_location++) {
        if (storedPublishes[store_location].packetId == 0) {
            break;
        }
    }

    if (store_location == MAX_PUBLISH_COUNT) {
        return false;
    }

    for (size_t i = 0; i < ioVecCount; i++) {
        memory_required += pIoVec[i].iov_len;
    }

    // The following is user dependent memory management function.
    // They can implement it however they want. They can simply use malloc
    // or get it from an array or anything else.
    uint8_t *buffer = user_memory_management_get_function(memory_required);
    size_t index = 0;

    for (size_t i = 0; i < ioVecCount; i++) {
        memcpy(&buffer[index], pIoVec[i].iov_base, pIoVec[i].iov_len);
        index += pIoVec[i].iov_len;
    }

    storedPublishes[store_location].buffer = buffer;
    storedPublishes[store_location].packetId = packetId;
    storedPublishes[store_location].bufferlen = memory_required;
    return true;
}

bool publishRetrieveCallback(struct MQTTContext* pContext,
                            uint16_t packetId,
                            uint8_t * buffer,
                            size_t * bufferLength)
{
    bool found = false;

    for ( int i = 0; i < MAX_PUBLISH_COUNT; i++ )
    {
        if (storedPublishes[i].packetId == packetId)
        {
            // No need to memcpy here. The buffer is on "user-heap" and stays there unless freed.
            buffer = storedPublishes[i].buffer;
            *bufferLength = storedPublishes[i].bufferlen;

            found = true;

            break;
        }
    }

    return found;
}

bool publishClearCallback(struct MQTTContext* pContext,
                          uint16_t packetId)
{
    bool found = false;

    for (int i = 0; i < MAX_PUBLISH_COUNT; i++)
    {
        if (storedPublishes[i].packetId == packetId)
        {
            // User is free to choose how to "free" this memory. They can simply call free or memset it to 0 or any other choice.
            user_memory_management_free_function(storedPublishes[i].buffer);
            storedPublishes[i].packetId = 0;
            found = true;
            break;
        }
    }

    return found;
}
```

Please Ignore the flouting of some MISRA rules in my example and inconsistent naming style :)

But, do go over the above and let me know if you think there are some flows with the system.

This is a blocker for me as we are now defining an API - the internal workings can change but I do not want to change any API often.

Thanks

File: source/core_mqtt_serializer.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (AniruddhaKanhere) [reply source/core_mqtt_serializer.c]: `if( a )` compiles to this assembly:
```asm
        ldrb    w0, [sp, 31]
        and     w0, w0, 255
        and     w0, w0, 1
        cmp     w0, 0        # This is the line to focus on.
        beq     .L2
```

and `if(a == true)` compiles to this:
```asm
        ldrb    w0, [sp, 31]
        and     w0, w0, 255
        and     w0, w0, 1
        cmp     w0, 0        # This is the line to focus on.
        beq     .L3
```

The only difference in both is the branch location - as it were two different branches.

**Review by AniruddhaKanhere — DISMISSED** [review]

File: .github/.cSpellWords.txt [hunk 7]
```diff
@@ -34,7 +36,10 @@ NONDET
 pylint
 pytest
 pyyaml
+serializemqttvec
 sinclude
+SHOU
```
Comment (archigup) [inline .github/.cSpellWords.txt:41]: Why is this needed?

File: .github/.cSpellWords.txt [hunk 8]
```diff
@@ -34,7 +36,10 @@ NONDET
...
+SHOULDNT
```
Comment (archigup) [inline .github/.cSpellWords.txt:42]: apostrophe?

File: .github/.cSpellWords.txt [hunk 9]
```diff
@@ -20,9 +20,11 @@ DLIBRARY
 DNDEBUG
 DUNITTEST
 DUNITY
+getbytesinmqttvec
 getpacketid
 isystem
 lcov
+LDNT
```
Comment (archigup) [inline .github/.cSpellWords.txt:27]: What is this?

File: .github/.cSpellWords.txt (hunk 8 above)
Comment (archigup) [reply .github/.cSpellWords.txt:42]: I dont see it used either

File: source/core_mqtt.c [hunk 10]
```diff
@@ -2681,6 +2784,53 @@ MQTTStatus_t MQTT_InitStatefulQoS( MQTTContext_t * pContext,
 
 /*-----------------------------------------------------------*/
 
+MQTTStatus_t MQTT_InitRetransmits( MQTTContext_t * pContext,
+                                   MQTTStorePacketForRetransmit storeFunction,
+                                   MQTTRetrievePacketForRetransmit retrieveFunction,
+                                   MQTTClearPacketForRetransmit clearFunction,
+                                   MQTTClearAllPacketsForRetransmit clearAllFunction )
```
Comment (archigup) [inline source/core_mqtt.c:2791]: Do we need clear all? or can we just call clear for all packetids we know?

File: .github/.cSpellWords.txt (hunk 8 above)
Comment (AniruddhaKanhere) [reply .github/.cSpellWords.txt:42]: It is used in a unit test string.

File: source/core_mqtt.c (hunk 10 above)
Comment (AniruddhaKanhere) [reply source/core_mqtt.c:2784]: For a clean session, we do need to clear all the memory allocated to publish messages. It is simpler to ask the user to clear all the memory instead of looping through the array one entry at a time. 
But, on the other hand, it is an additional function that the user has to provide.

I can see the value in both approaches. But, I think that asking the user to provide one more function is bad for user experience. To that end, I think we can remove this.

Comment (AniruddhaKanhere) [reply source/core_mqtt.c:2784]: Thinking a bit more, having a clear all function seems useful. If the user is using an array - they can just memset that to 0.
Thoughts @archigup?

File: source/core_mqtt_serializer.c [hunk 11]
```diff
@@ -2623,6 +2628,33 @@ MQTTStatus_t MQTT_GetIncomingPacketTypeAndLength( TransportRecv_t readFunc,
...
+MQTTStatus_t MQTT_UpdateDuplicatePublishFlag( uint8_t * pHeader,
+                                              bool set )
+{
+    MQTTStatus_t status = MQTTSuccess;
+
+    if( pHeader == NULL )
+    {
+        status = MQTTBadParameter;
+    }
+    else if( ( ( *pHeader ) & MQTT_PACKET_TYPE_PUBLISH ) == 0 )
```
Comment (archigup) [inline source/core_mqtt_serializer.c:2640]: Is this correct? Should it not be `((*pHeader) & 0xF0U) != MQTT_PACKET_TYPE_PUBLISH`?

File: source/include/core_mqtt.h [hunk 12]
```diff
@@ -101,6 +107,61 @@ typedef void (* MQTTEventCallback_t )( struct MQTTContext * pContext,
...
+ *                MQTTVec_t in the provided memory respectively.
+ * @param[in] mqttVecCount Number of transport vectors in the pIoVec array.
+ *
+ * @return True if the copy is successful else false.
+ */
+/* @[define_mqtt_retransmitstorepacket] */
+typedef bool ( * MQTTStorePacketForRetransmit)( struct MQTTContext * pContext,
+                                                uint16_t packetId,
+                                                MQTTVec_t * pMqttVec,
+                                                size_t mqttVecCount );
```
Comment (archigup) [inline source/include/core_mqtt.h:127]: shouldn't mqttVecCount be part of the MQTTVec_t? Seems like leaking an internal detail

File: source/core_mqtt_serializer.c (hunk 11 above)
Comment (AniruddhaKanhere) [reply source/core_mqtt_serializer.c:2640]: We are just verifying whether the publish bit is set. Other bits can be set as well.

I think this should suffice.

File: source/include/core_mqtt.h (hunk 12 above)
Comment (AniruddhaKanhere) [reply source/include/core_mqtt.h]: Fixed

**Review by archigup — APPROVED** [review]

**Review by AniruddhaKanhere — APPROVED** [review]
